src/inset_state/densify.cpp

```diff
@@ -3,33 +3,74 @@
-  // either way.
-  // Check if any segment is undefined (i.e., defined by identical points)
-  // if (a1 == a2 || b1 == b2) {
-  //   std::cerr << "ERROR: End points of line segment are identical"
-  //             << std::endl;
-  //   _Exit(EXIT_FAILURE);
-  // }
-
-  // Return segment intersection if it exists
-  const auto result = CGAL::intersection(s1, s2);
+// This function takes the following as input:
+// - a segment defined by points a and b.
+// - a line defined by coef_x, coef_y, and intercept with the formula
+//      coef_x * x + coef_y * y + intercept = 0.
+// The function returns the unique intersection point between them. If this
+// intersection point does not exist, the function returns the point (-1, -1),
+// which is always outside of any graticule grid cell.
+Point calc_intersection(const Point a,
+                        const Point b,
+                        const double coef_x,
+                        const double coef_y,
+                        const double intercept,
+                        const unsigned int lx,
+                        const unsigned int ly)
+{
+  const auto result = CGAL::intersection(
+    Line(coef_x, coef_y, intercept),
+    Segment(a, b)
+  );
   if (result) {
     if (const Point* p = boost::get<Point>(&*result)) {
-      return (*p);
+      return rounded_point((*p), lx, ly);
     }
   }
+  return Point(-1.0, -1.0);
+}
 
-  // Value outside grid cell indicating that no intersection was found
-  return Point(-1, -1);
+// This function takes as input points a and b which define a segment, and
+// the following arguments which defines the kind of diagonals for which we
+// will calculate intersections:
+// - slope: the slope of every diagonal.
+// - base_intercept: the intercept that is the closest to 0 of a diagonal
+// on the grid.
+// - step: what we need to add to each diagonal's intercept to obtain the
+// next diagonal.
+// - edge: a char, which can be either 'x', 'y', or 'n'. 'x' marks diagonals
+// on the left or right edges of the grid, 'y' marks diagonals on the top or
+// bottom edges, and 'n' marks non-edge diagonals.
+// TODO: we can also determine whether it's an edge diagonal or not just by
+// looking at the slope, but having a separate edge argument might be clearer.
+// What should we do?
+void add_diag_inter(std::set<Point, decltype(point_lesser)*>
+                      *intersections,
+                    const Point a,
+                    const Point b,
+                    double slope,
+                    double base_intercept,
+                    double step,
+                    const unsigned int lx,
+                    const unsigned int ly,
+                    char edge)
+{
+  double intercept_start =
+    floor(std::min(a.y() - slope * a.x(), b.y() - slope * b.x()))
+    + base_intercept;
+  double intercept_end =
+    std::max(a.y() - slope * a.x(), b.y() - slope * b.x());
+  for (double i = intercept_start; i <= intercept_end; i += step) {
```

Comment:
As commented above, I do not think that `i` is an intercept. Can we choose another variable name? How about `d` for double?